packages/backend-common/src/reading/AzureUrlReader.ts

```diff
@@ -47,14 +48,19 @@ import { ReadUrlResponseFactory } from './ReadUrlResponseFactory';
 export class AzureUrlReader implements UrlReader {
   static factory: ReaderFactory = ({ config, treeResponseFactory }) => {
     const integrations = ScmIntegrations.fromConfig(config);
+    const provider =
+      DefaultAzureDevOpsCredentialProvider.fromIntegrations(integrations);
     return integrations.azure.list().map(integration => {
-      const reader = new AzureUrlReader(integration, { treeResponseFactory });
+      const reader = new AzureUrlReader(provider, integration, {
+        treeResponseFactory,
+      });
       const predicate = (url: URL) => url.host === integration.config.host;
       return { reader, predicate };
     });
   };
 
   constructor(
+    private readonly provider: AzureDevOpsCredentialsProvider,
     private readonly integration: AzureIntegration,
     private readonly deps: { treeResponseFactory: ReadTreeResponseFactory },
```

Comment:
Let's pass the `provider` through `deps` instead. I'm feeling it's best to call it `credentialsProvider` too, to keep that clearer in the implementation

Reply:
@Rugvip, what do you mean by this?
> let's pass the provider through deps instead

Should I modify the `ReaderFactory` to something like this?
```typescript
static factory: ReaderFactory = ({ config, treeResponseFactory, credentialsProvider }) => { ... }
```

From what I can tell, that, in turn, would mean that the `UrlReadersOptions` also needs an `azureCredentialsProvider` property; then the `azureCredentialsProvider` is provided as part of `makeCreateEnv`?
```typescript
function makeCreateEnv(config: Config) {
  const root = getRootLogger();
  const reader = UrlReaders.default({ logger: root, config, azureCredendialsProvider });
  //...
```

Since the provider caches tokens, we could go with `getAzureCredentialsProvider` and `seAzureCredentialsProvider` functions similar to `getRootLogger()` and `setRootLogger()`.

Or do you mean that the `AzureCredentialsProvider` should become part of the [core services](https://backstage.io/docs/backend-system/architecture/services/#core-services)? 

I checked how this is done for the `DefaultGithubCredentialsProvider`, and that provider also instantiated inside the `factory` method:
```typescript
export class GithubUrlReader implements UrlReader {
  static factory: ReaderFactory = ({ config, treeResponseFactory }) => {
    const integrations = ScmIntegrations.fromConfig(config);
    const credentialsProvider =
      DefaultGithubCredentialsProvider.fromIntegrations(integrations);
    return integrations.github.list().map(integration => {
      const reader = new GithubUrlReader(integration, {
        treeResponseFactory,
        credentialsProvider,
      });
      //...
```
 https://github.com/backstage/backstage/blob/698b0d749b9c5885c39866a0569663f799b0892a/packages/backend-common/src/reading/GithubUrlReader.ts#LL64C5-L65C71

Reply:
I didn't read your comment carefully enough 😓. I modified the code to pass the `credentialsProviders` through the `deps` object.